### PR TITLE
Initial working version of a playground

### PR DESCRIPTION
**Pull request by emptycrown**
Working MVP that enables testing a list of indices across multiple modes and comparing results. In the future, we can add more things to test over (for example, text splitters!) but this is a solid start. Already useful in gaining insight into the different indices, modes, LLMs, embeddings, etc.

### CODE REVIEW

File: examples/playground/.modules/library.json [hunk 1]
```diff
@@ -0,0 +1,164 @@
+{
```
Comment (jerryjliu) [inline examples/playground/.modules/library.json]: wait why are we checking in .modules, let's let the user download this

File: gpt_index/__init__.py [hunk 2]
```diff
@@ -42,6 +42,9 @@
 from gpt_index.langchain_helpers.memory_wrapper import GPTIndexMemory
 from gpt_index.langchain_helpers.sql_wrapper import SQLDatabase
 
+# playground
+from gpt_index.playground.base import Playground
```
Comment (jerryjliu) [inline gpt_index/__init__.py]: i kinda effed up by dumping everything in the root init file, now import times for gpt index are already really slow. if you don't mind let's remove Playground here so that users have to do `from gpt_index.playground import Playground`

File: gpt_index/playground/base.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+class Playground:
+    """Experiment with indices, model, embeddings, modes, and more."""
+
+    def __init__(
+        self, indices: List[BaseGPTIndex], modes: Optional[List[str]] = DEFAULT_MODES
+    ):
+        """Initialize with indices to experiment with.
+
+        Args:
+            indices: A List of BaseGPTIndex's to experiment with
```
Comment (jerryjliu) [inline gpt_index/playground/base.py]: document modes as well (there's no direct reference to the QueryMode class in the docs but the modes are linked under each query class here: https://gpt-index.readthedocs.io/en/latest/reference/query.html).

File: gpt_index/playground/base.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+DEFAULT_INDEX_CLASSES = [GPTSimpleVectorIndex, GPTTreeIndex, GPTListIndex]
+DEFAULT_MODES = ["default", "summarize", "embedding", "retrieve", "recursive"]
+
+
+class Playground:
+    """Experiment with indices, model, embeddings, modes, and more."""
+
+    def __init__(
+        self, indices: List[BaseGPTIndex], modes: Optional[List[str]] = DEFAULT_MODES
```
Comment (jerryjliu) [inline gpt_index/playground/base.py]: why do indices not have a default here?

File: gpt_index/playground/base.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        """Compare index outputs on an input query.
+
+        Args:
+            query_text (str): Query to run all indices on.
+            to_pandas (Optional[bool]): Return results in a pandas dataframe. True by default.
+
+        Returns:
+            The output of each index along with other data, such as the time it took to compute. Results are stored in a Pandas Dataframe or a list of Dicts.
+        """
+        print(f"\033[1mQuery:\033[0m\n{query_text}\n")
```
Comment (jerryjliu) [inline gpt_index/playground/base.py:115]: was going to suggest using the logging library, but not sure if that plays nicely with your color formatting

File: gpt_index/playground/base.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    def __init__(
+        self, indices: List[BaseGPTIndex], modes: Optional[List[str]] = DEFAULT_MODES
+    ):
+        """Initialize with indices to experiment with.
+
+        Args:
+            indices: A List of BaseGPTIndex's to experiment with
+        """
+        self.update_indices(indices)
```
Comment (jerryjliu) [inline gpt_index/playground/base.py]: the term `update` is a big confusing, think for this case it's more canonical to define a setter https://www.geeksforgeeks.org/getter-and-setter-in-python/. You can abstract out the validation logic in a `_validate_indices` method (same with modes)

**Review by jerryjliu — COMMENTED** [review]
this is nice! could we add a test or two for this? (like at the minimum to make sure it runs without errors, even without assertions)

**Comment by jerryjliu** [comment]
also would love to document this in the docs somewhere (but we can add later)

File: examples/playground/.modules/library.json (hunk 1 above)
Comment (emptycrown) [reply examples/playground/.modules/library.json]: Oh oops, thought I deleted this

File: gpt_index/playground/base.py (hunk 4 above)
Comment (emptycrown) [reply gpt_index/playground/base.py]: These are objects that already have documents in them. Documents are specific to the user so doesn't make sense to have default here. However, I made a default option through the `from_docs` method.

File: gpt_index/playground/base.py (hunk 6 above)
Comment (emptycrown) [reply gpt_index/playground/base.py]: Oh cool

**Comment by emptycrown** [comment]
Wrote tests. Documentation sounds good. Does `readthedocs` update automatically based on the `docs` directory?

**Comment by jerryjliu** [comment]
yep! you can check for yourself by going into `docs` folder and typing `make html`

File: gpt_index/playground/base.py (hunk 4 above)
Comment (jerryjliu) [reply gpt_index/playground/base.py]: Oooooh right oops i thought this was default index classes. Actually shouldn't index_classes be specifiable by the user?

Comment (emptycrown) [reply gpt_index/playground/base.py]: Hm interesting. That's probably right -- I was thinking that if you were going to specify indices, you might as well just initialize normally. I'll add this in though.

**Review by jerryjliu — APPROVED** [review]
🎉
